bumble/drivers/intel.py

```diff
@@ -0,0 +1,343 @@
+
+
+class Driver:
+    def __init__(self, host, version: IntelVersionTlv, firmware: bytes):
+        self.host = host
+        self.version = version
+        self.firmware = firmware
+
+    @classmethod
+    async def for_host(cls, host):  # type: ignore
+        try:
+            version = await fetch_intel_version(host)  # type: ignore
+            fw = prepare_firmware(version)
+            return cls(host, version, fw)
+        except Exception as e:
+            logging.error(f"Error preparing the firmware {e}")
+            return None
+
+    async def init_controller(self):
+        try:
+            await download_firmware(self.host, self.version, self.firmware)
+        except Exception as e:
+            logging.error(f"Failed to download the firmware: {e}")
+            return None
+
+
+async def fetch_intel_version(host) -> IntelVersionTlv:  # type: ignore
+    response = await host.send_command(HCI_Intel_Read_Version_Command(param=0xFF), check_result=True)  # type: ignore
+    if response.return_parameters.status != HCI_SUCCESS:  # type: ignore
+        raise ValueError("This controller is not an intel device")
+
+    if response.return_parameters.data[0] == 0x37:  # type: ignore
+        raise ValueError("Legacy Intel unsupported")
+
+    intel_version_tlvs = parse_intel_version_tlvs(response.return_parameters.data)  # type: ignore
+
+    if intel_version_tlvs.cnvi_bt is None:
+        raise ValueError("CNVI_BT cannot be None")
+
+    intel_hw_platform = intel_version_tlvs.cnvi_bt.value[1]
+    if intel_hw_platform != HW_PLATFORM:
+        raise ValueError("Unsupported Intel hardware platform")
+
+    intel_hw_variant = (
+        int.from_bytes(intel_version_tlvs.cnvi_bt.value, 'little') & 0x003F0000
+    ) >> 16
+    match intel_hw_variant:
+        case 0x17 | 0x18 | 0x19 | 0x1B | 0x1C:
+            return intel_version_tlvs
+        case _:
+            raise ValueError("Unsupported Intel hardware variant")
+
+
+def intel_cnvx_top_pack_swab(top: int, step: int) -> int:
+    combined: int = ((top << 4) | step) & 0xFFFF
+
+    return ((combined >> 8) & 0xFF) | ((combined & 0xFF) << 8)
+
+
+def fetch_firmware_name(version: IntelVersionTlv) -> str:
+    if version.cnvi_top is None or version.cnvr_top is None:
+        raise ValueError("cnvi_top and cnvr_top cannot be None")
+
+    cnvi_top = int.from_bytes(version.cnvi_top.value, byteorder='little')
+    cnvi_top_step = (cnvi_top & 0x0F000000) >> 24
+    cnvi_top_type = cnvi_top & 0x00000FFF
+
+    cnvr_top = int.from_bytes(version.cnvr_top.value, byteorder='little')
+    cnvr_top_step = (cnvr_top & 0x0F000000) >> 24
+    cnvr_top_type = cnvr_top & 0x00000FFF
+
+    upper_name = intel_cnvx_top_pack_swab(cnvi_top_type, cnvi_top_step)
+    lower_name = intel_cnvx_top_pack_swab(cnvr_top_type, cnvr_top_step)
+
+    return f"ibt-{upper_name:04x}-{lower_name:04x}.sfi"
+
+
+def prepare_firmware(version: IntelVersionTlv) -> bytes:
+    if not sys.platform.startswith("linux"):
```

Comment:
It would be nice to be able to load FW images from other places than the default Linux location, like for the Realtek driver (see rtk.py `Driver.find_binary_path`)